### PR TITLE
Swagger definition for Get device location

### PR DESCRIPTION
**Pull request by bigludo7**
First stab for API definition for **retrieving a device location**. There are points to be commented in this PR and discussed in team meeting:

- I've not yet provided any documentation - will do once we agreed on the interface directly in the OAS.
- Use of device identifiers as defined in latest QoD version (phoneNumber an not MSISDN) - this introduce a gap with the check api.
- Definition of a specific scope distinct from the check operation
- In the response only lat/log & accuracy are provided. I opted to provided straightforward information and not use area schema (with oneOf) --> could be challenged.
- We need some update on the other api - for example the little for the existing api should be check device location API instead of device location API. 
- etc.

### CODE REVIEW

File: code/API_definitions/Get DeviceLocation.yaml
Comment (jlurien) [inline code/API_definitions/Get DeviceLocation.yaml:38]: Here we have to decide whether we model this as GET or as POST. GET is more natural options to retrieve data but it has the problem of modelling complex input parameters. If we wanted to reuse the same `device` object that we have in QoD or the original spec of Location, we could pass an object as query parameter, which is valid, but it has to be encoded properly, which is less readable. Reusing the object as it is has the advantages that it is possible to express the constraints between the different properties, for example that port is required along with public IP address.
Modelling as POST is also possible and align with other APIs but guidelines are not very clear about this. We could model the endpoint as POST /retrieve-location, so options would be:

UrlEncoded( GET location/v0/locations?device={"ipV4Address":{"public_ip_address":"1.2.3.4","public_port":1234}})

or

POST /location/v0/retrieve-location
{
  "device": {
    "ipV4Address": {
      "public_ip_address": "1.2.3.4",
      "public_port": 1234
    }
  }
)

There are similar cases in other APIs so maybe it makes sense to raise this in commonalities and adjust guidelines accordingly

File: code/API_definitions/Get DeviceLocation.yaml
Comment (jlurien) [inline code/API_definitions/Get DeviceLocation.yaml:143]: This is the most obvious response. Not sure if in some cases a different Area other than Circle would make sense, if network can return a different geometry.

**Review by jlurien — COMMENTED** [review]
Thanks for providing first approach. My main doubt is how to model data retrieval with complex inputs, GET vs POST, a recurrent debate while designing APIs :)

File: code/API_definitions/Get DeviceLocation.yaml
Comment (bigludo7) [reply code/API_definitions/Get DeviceLocation.yaml:28]: Yes agreed - When I crafted this proposal I've also feeling that we need to update the check location API accordingly.
Pragmatically do you think, @jlurien , that I should do this in same PR ?

I tend to think yes but looking for guidance.

File: code/API_definitions/Get DeviceLocation.yaml
Comment (bigludo7) [reply code/API_definitions/Get DeviceLocation.yaml:31]: Yes agrees - note this is related to previous point as existing check location is a POST` /localisations `with a s. (but only one device checked). i have keep consistency.
My preference is also to have both endpoints in singular.

File: code/API_definitions/Get DeviceLocation.yaml
Comment (bigludo7) [reply code/API_definitions/Get DeviceLocation.yaml:143]: Same for me. I guess circle is well enough for first release. 

Let's wait for UC requiring other Area definition and in this case we'll implement the `oneOf `model.

File: code/API_definitions/Get DeviceLocation.yaml
Comment (bigludo7) [reply code/API_definitions/Get DeviceLocation.yaml:38]: Yes this is a point to be discussed. 
Looking in other API we have POST in device status right now and GET for Device-identifier

i will raise this point in commonalities but not sure we'll get a rule.

Regarding our case, I will probably prefer to use POST pattern to align with Check. 

Let's tab this point for next meeting.

**Comment by bigludo7** [comment]
Discussion in Commonalities --> https://github.com/camaraproject/WorkingGroups/issues/212

**Review by bigludo7 — COMMENTED** [review]
Update after first set of review.
Mainly changed GET operation to POST

**Review by bigludo7 — COMMENTED** [review]
Renamed resource retrieve-location (instead of location) as using POST operation

File: code/API_definitions/Get DeviceLocation.yaml
Comment (JoachimDahlgren) [reply code/API_definitions/Get DeviceLocation.yaml]: Agree that POST is probably a better way as the input is complex and it aligns with both device status as well as with verify location.

File: code/API_definitions/Get DeviceLocation.yaml
Comment (JoachimDahlgren) [reply code/API_definitions/Get DeviceLocation.yaml:162]: Point is probably good as a starting point. There is however a potential complexity in providing only a point. Depending on the southbound system a request for location might come in another format, such as a polygon,  which then  needs to be translated into a point with an appropriate accuracy.

File: code/API_definitions/Get DeviceLocation.yaml
Comment (JoachimDahlgren) [inline code/API_definitions/Get DeviceLocation.yaml:97]: Should the UeId be renamed to Device to be aligned with QoD 0.9?

File: code/API_definitions/Get DeviceLocation.yaml
Comment (JoachimDahlgren) [inline code/API_definitions/Get DeviceLocation.yaml:103]: UeId should probably be aligned with QoD 0.9. Be renamed to Device with properties phoneNumber, networkAccessIdentifier, ipv4Address, ipv6Address with Port moving to the ipv4Address as publicPort.

Comment (bigludo7) [reply code/API_definitions/Get DeviceLocation.yaml:103]: This is a fair comment. I've aligned with our existing check location API but probably we should align both device location api with Qod v0.9 if this later is enough mature.

**Review by bigludo7 — COMMENTED** [review]
Alignement with QoD device identification structure.

**Review by bigludo7 — COMMENTED** [review]
Alignement with QoD device identification structure.

**Comment by bigludo7** [comment]
@JoachimDahlgren @jlurien @rartych

- changed operation to POST
- aligned with QoD device identifier structure.

I suggest to tackle check-device device identifier structure alignment in another PR once this one merged.
Then I will create 2 other PR to add subscription management.

File: code/API_definitions/Get DeviceLocation.yaml
Comment (jlurien) [reply code/API_definitions/Get DeviceLocation.yaml:28]: For decisions which affect both APIs I think it is ok to make changes in same PR

File: code/API_definitions/Get DeviceLocation.yaml
Comment (jlurien) [inline code/API_definitions/Get DeviceLocation.yaml:157]: This schema is not used

**Review by bigludo7 — COMMENTED** [review]
Fixed issues identified by @jlurien

File: code/API_definitions/Get DeviceLocation.yaml
Comment (bigludo7) [reply code/API_definitions/Get DeviceLocation.yaml]: Yes correct _ fixed

File: code/API_definitions/Get DeviceLocation.yaml
Comment (jlurien) [reply code/API_definitions/Get DeviceLocation.yaml:28]: Thought better of it, let's make separate PRs for changes in  the other API. There is already one to adapt ueId to device.

File: code/API_definitions/Get DeviceLocation.yaml
Comment (akoshunyadi) [inline code/API_definitions/Get DeviceLocation.yaml:31]: I have to note here, that our guidelines says: "Verbs use is not allowed."
Maybe: location-detection or location-retrieval

File: code/API_definitions/Get DeviceLocation.yaml
Comment (akoshunyadi) [reply code/API_definitions/Get DeviceLocation.yaml:186]: Shouldn't we add a time info of this location data? I think this will be needed sooner or later, since it can be quite important for the clients.

File: code/API_definitions/Get DeviceLocation.yaml
Comment (akoshunyadi) [reply code/API_definitions/Get DeviceLocation.yaml:31]: @jlurien ok, then it's fine for me ( in the location API there was also a 'verify' initially )

File: code/API_definitions/Get DeviceLocation.yaml
Comment (bigludo7) [reply code/API_definitions/Get DeviceLocation.yaml:186]: @akoshunyadi good question - For me it depends on if we want the API to provides current location or not last known location (associated to an age info).  This is tricky because depending on the business case one or the other could be required. 
I was thinking to keep thing simple and just provide current location or 404 if not able to provide but open to discuss this (add for example an `ageOfLocation `attribute in the response)

@jlurien WDYT?

**Review by bigludo7 — COMMENTED** [review]
Applied decisions done by the team June 21th:
Change accuracy unit (from km to meters)
Add maxAge in request (in seconds - min 60)
Add ageOfLocation in response

**Comment by bigludo7** [comment]
Hello @jlurien @akoshunyadi @JoachimDahlgren @rartych 
Kind reminder
I've applied all required changes 2 weeks ago. 
Looking for your review as we planned to release this API in July
Thanks

**Review by akoshunyadi — APPROVED** [review]

File: code/API_definitions/Get DeviceLocation.yaml
Comment (jlurien) [inline code/API_definitions/Get DeviceLocation.yaml:211]:
```suggestion
          description: Last date and time when the device was localized (using ISO 8601 with timezone)
```

File: code/API_definitions/Get DeviceLocation.yaml
Comment (jlurien) [inline code/API_definitions/Get DeviceLocation.yaml:209]:
```suggestion
          description: Provided accuracy for the location, in meters, from the point defined by latitude/longitude

```

File: code/API_definitions/Get DeviceLocation.yaml
Comment (jlurien) [inline code/API_definitions/Get DeviceLocation.yaml:103]:
```suggestion
          description: Maximum age of the location information which is accepted for the location retrieval (in seconds)

```

File: code/API_definitions/Get DeviceLocation.yaml
Comment (jlurien) [inline code/API_definitions/Get DeviceLocation.yaml:230]: We should define an error when the `maxAge` threshold cannot be satisfied.

**Review by jlurien — COMMENTED** [review]
Some suggestions

File: code/API_definitions/Get DeviceLocation.yaml
Comment (bigludo7) [reply code/API_definitions/Get DeviceLocation.yaml:4]: Yes - will do once we validate the OAS.

**Review by bigludo7 — COMMENTED** [review]
Following review:
- Changed scope name
- Added specific error for maxAge
- Integrated inheritance mode for retrieved area

**Comment by bigludo7** [comment]
@jlurien : updated the OAS - looking for your feedback in particular for the integration of the inheritance model for the area.

**Review by bigludo7 — COMMENTED** [review]
fixed example typo

**Comment by bigludo7** [comment]
@jlurien Everything look resolved from my perspective - Ok for you? Waiting for your approval to merge it

**Comment by jlurien** [comment]
> @jlurien Everything look resolved from my perspective - Ok for you? Waiting for your approval to merge it

Let's discuss it tomorrow. Merging it now may impact the releasing process for v0.2. We may prefer to wait until v0.2 for location-verification is released and then merge the first PRs for the new APIs in the family.

File: code/API_definitions/Get DeviceLocation.yaml
Comment (jlurien) [inline code/API_definitions/Get DeviceLocation.yaml:97]:
```suggestion
            location-retrieval-read: Provide device location
```

**Review by jlurien — APPROVED** [review]
Mostly OK. I have some comments to align "aesthetically" both location-verification and this API, in terms of title, path name (verb vs plural name), scope (it is ok in security but not in securitySchemes)

**Review by bigludo7 — COMMENTED** [review]
Updated following @jlurien review.

File: code/API_definitions/Get DeviceLocation.yaml
Comment (bigludo7) [reply code/API_definitions/Get DeviceLocation.yaml:80]: Done !

**Review by jlurien — DISMISSED** [review]
LGTM

File: code/API_definitions/Get DeviceLocation.yaml
Comment (jlurien) [inline code/API_definitions/Get DeviceLocation.yaml:29]:
```suggestion
  /retrieve:
```
Agreed after #84, to consolidate naming in both APIs

**Review by jlurien — COMMENTED** [review]
Rename path to just /retrive, as location-verification has been renamed to /verify

**Review by bigludo7 — COMMENTED** [review]
Changed resource name to /retrieve

**Review by bigludo7 — COMMENTED** [review]
Alignment for oAuth2AuthorizationCode & scope as done in PR #88

File: code/API_definitions/Get DeviceLocation.yaml
Comment (jlurien) [inline code/API_definitions/Get DeviceLocation.yaml:18]: If the API name is just "location", then scope would be "location:retrieve"

File: code/API_definitions/Get DeviceLocation.yaml
Comment (jlurien) [inline code/API_definitions/Get DeviceLocation.yaml:95]: location:retrieve

File: code/API_definitions/Get DeviceLocation.yaml
Comment (jlurien) [inline code/API_definitions/Get DeviceLocation.yaml:293]: If we follow also API_NAME.{CODE}  then it would be just LOCATION.MAXAGE_INVALID_ARGUMENT

**Review by jlurien — COMMENTED** [review]
All comments around API name, depending on whether it is just "location" or is "location-retrieval". It affects scopes and error codes naming. If we choose "location-retrieval" then servers must be updated.

**Review by bigludo7 — COMMENTED** [review]
changer server name to `location-retrieval`

**Review by jlurien — APPROVED** [review]
I think it is ready to merge as v0.1-wip

**Comment by bigludo7** [comment]
@jlurien Agreed.
I will wait till Monday and then proceed (and work on the documentation part)
As it is still in wip we can still accommodate comment after - if any.

**Pull request merged**
